### PR TITLE
fix(trie): discard zero-value slots in hashed poststate cursor

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Issue

The in-memory `PostState` cursor currently returns zero-value storage slots which shouldn't be part of the state trie

## Solution

Discard zeroed-out entries from both in-memory & database cursors.

### CODE REVIEW

File: crates/trie/src/hashed_cursor/post_state.rs [hunk 1]
```diff
@@ -210,13 +210,24 @@ pub struct HashedPostStateStorageCursor<'b, C> {
...
+    fn is_db_storage_wiped(&self, account: &H256) -> bool {
         match self.post_state.storages.get(account) {
             Some(storage) => storage.wiped,
             None => false,
         }
     }
 
+    /// Check if the slot was zeroed out in the post state.
+    /// The database is not checked since we don't insert zero valued slots.
+    fn is_slot_empty(&self, account: &H256, slot: &H256) -> bool {
```
Comment (mattsse) [inline crates/trie/src/hashed_cursor/post_state.rs:222]: this returns false if no entry exists, is this expected?

should this be called `is_(touched?)_slot_value_zero`?

**Review by mattsse — CHANGES_REQUESTED** [review]
not really familiar with these rules, but I think `empty` can be a bit ambiguous.

File: crates/trie/src/hashed_cursor/post_state.rs (hunk 1 above)
Comment (rkrasiuk) [reply crates/trie/src/hashed_cursor/post_state.rs:222]: that is indeed expected. `is_touched_slot_value_zero` sounds much clearer

**Review by mattsse — APPROVED** [review]

**Review by onbjerg — APPROVED** [review]
looks good to me, although i am not sure if empty storage slots are supposed to be part of the trie for accounts pre state clearing :shrug:

**Pull request merged**